go/vt/vttablet/grpctmclient/cached_client.go

```diff
@@ -0,0 +1,589 @@
+		return dial(addr)
+	}
+
+	select {
+	// Non-blocking signal to the janitor that it should consider sweeping soon.
+	case client.janitor.ch <- sentinel:
+	default:
+	}
+
+	if !client.sema.AcquireContext(ctx) {
+		return nil, nil, ctx.Err()
+	}
+
+	return dial(addr)
+}
+
+func (client *cachedClient) Close() {
+	client.m.Lock()
+	defer client.m.Unlock()
+
+	close(client.janitor.ch)
+	client.sweepFnLocked(func(conn *pooledTMC) (bool, bool) {
+		return true, false
+	})
+}
+
+func (client *cachedClient) sweep2(f func(conn *pooledTMC) (bool, bool)) {
+	client.m.RLock()
+
+	var toFree []string
+	for key, conn := range client.conns {
+		conn.m.RLock()
+		shouldFree, stopSweep := f(conn)
+		conn.m.RUnlock()
+
+		if shouldFree {
+			toFree = append(toFree, key)
+		}
+
+		if stopSweep {
+			break
+		}
+	}
+
+	client.m.RUnlock()
+
+	if len(toFree) > 0 {
+		client.m.Lock()
+		defer client.m.Unlock()
+
+		for _, key := range toFree {
+			conn, ok := client.conns[key]
+			if !ok {
+				continue
+			}
+
+			conn.m.Lock()
+			// check the condition again, things may have changed since we
+			// transitioned from the read lock to the write lock
+			shouldFree, _ := f(conn)
+			if !shouldFree {
+				conn.m.Unlock()
+				continue
+			}
+
+			conn.cc.Close()
+			conn.m.Unlock()
+			delete(client.conns, key)
+			client.sema.Release()
+		}
+	}
+}
+
+func (client *cachedClient) sweep() {
+	now := time.Now()
+	client.sweep2(func(conn *pooledTMC) (bool, bool) {
+		return conn.refs == 0 && conn.lastAccessTime.Add(client.idleTimeout).Before(now), false
+	})
+}
+
+func (client *cachedClient) sweepFnLocked(f func(conn *pooledTMC) (shouldFree bool, stopSweep bool)) {
+	for key, conn := range client.conns {
+		conn.m.Lock()
+		shouldFree, stopSweep := f(conn)
+		if !shouldFree {
+			conn.m.Unlock()
+			if stopSweep {
+				return
+			}
+
+			continue
+		}
+
+		conn.cc.Close()
+		delete(client.conns, key)
+		client.sema.Release()
+		conn.m.Unlock()
+
+		if stopSweep {
+			return
+		}
+	}
+}
+
+var sentinel = &struct{}{}
+
+type janitor struct {
+	ch     chan *struct{}
+	client *cachedClient
+	timer  *timer.Timer
+
+	m         sync.Mutex
+	sweeping  bool
+	lastSweep time.Time
+}
+
+func (j *janitor) run() {
+	j.timer.Start(j.sweep)
+	defer j.timer.Stop()
+
+	for s := range j.ch {
+		if s == nil {
+			break
+		}
+
+		scan := true
+		t := time.NewTimer(time.Millisecond * 50) // TODO: flag
+		for scan {
+			select {
+			case <-t.C:
+				scan = false
+			case s := <-j.ch:
+				if s == nil {
+					scan = false
+				}
+			default:
+				scan = false
+			}
+		}
+
+		t.Stop()
+		j.sweep()
+	}
+}
+
+func (j *janitor) sweep() {
+	j.m.Lock()
+	if j.sweeping {
+		j.m.Unlock()
+		return
+	}
+
+	if j.lastSweep.Add(time.Millisecond * 10 /* TODO: flag */).After(time.Now()) {
+		j.m.Unlock()
+		return
+	}
+
+	j.sweeping = true
+	j.m.Unlock()
+
+	j.client.sweep()
+	j.m.Lock()
+	j.sweeping = false
+	j.lastSweep = time.Now()
+	j.m.Unlock()
+}
+
+type cachedConn struct {
```

Comment:
Relevant section I'm planning to test starts here. Everything above is an older implementation that got really slow on pool sizes above like 100, because finding old connections was O(n) and locked the cache for that entire time